### PR TITLE
factors out s3 store into core

### PR DESCRIPTION
**Pull request by pretentious7**

### CODE REVIEW

File: crates/y-sweet-core/src/store/s3.rs [hunk 1]
```diff
@@ -1,17 +1,27 @@
...
+    pub bucket: String,
+    pub region: String,
+    pub bucket_prefix: Option<String>,
+}
 
 const PRESIGNED_URL_DURATION_SECONDS: u64 = 60 * 60;
 pub struct S3Store {
     bucket: Bucket,
-    _bucket_inited: RefCell<bool>,
+    _bucket_inited: RwLock<bool>,
```
Comment (paulgb) [inline crates/y-sweet-core/src/store/s3.rs:24]: can this just be an `AtomicBool`?

**Review by paulgb — COMMENTED** [review]
I like it! Not done yet but a few comments for now.

I'm still not sure I understand the value of checking if the bucket exists on first use -- is it because otherwise we can't tell whether a 404 means that no bucket exists or no key exists?

**Comment by pretentious7** [comment]
https://github.com/drifting-in-space/y-sweet/pull/138

> I like it! Not done yet but a few comments for now.
> 
> I'm still not sure I understand the value of checking if the bucket exists on first use -- is it because otherwise we can't tell whether a 404 means that no bucket exists or no key exists?

For get(), yes, that's correct. But mostly I wanted the Store trait functions to be able to assume the bucket exists, or error out.  The first use thing is just so that S3Store::new is not async.

File: crates/y-sweet-core/src/store/s3.rs (hunk 1 above)
Comment (pretentious7) [reply crates/y-sweet-core/src/store/s3.rs:24]: yep, I've use SeqCst ordering cause errors should be more easily reproducible.

**Review by paulgb — APPROVED** [review]

**Pull request merged**